### ISSUE TITLE
Lua errors and segmentation fault when using Lua opcode example

### ISSUE THREAD

**Issue by allen-marshall** [csound/csound#959]
I am having an issue getting Lua opcodes to work in Csound on Ubuntu 17.10 (64-bit). Whenever I try to render code that uses a Lua opcode, including the example at [https://csound.com/docs/manual/lua.html](https://csound.com/docs/manual/lua.html), Csound generates a large number of Lua errors saying "attempt to call a nil value", then I get a segmentation fault message and Csound hangs until terminated with Ctrl+C. I am getting this behavior on both the develop branch and the 6.11.0 tag.

I have tried with the original example, and with `ffi.load('csound64.dll.5.2')` changed to `ffi.load('csound64')` so that Lua can find the shared library on my system. Without that change, I get an error message `luaL_dostring failed with: 1` that is not present with the change, presumably because the original cannot find libcsound64.so. In both cases, however, I get the Lua errors and segmentation fault after the orchestra initialization.

Here is the output that I get from running `csound luamoog-original.csd` (on the original example file) and `csound luamoog-modified.csd` (on the modified example file):

[csound-output-luamoog-original.txt](https://github.com/csound/csound/files/1996960/csound-output-luamoog-original.txt)
[csound-output-luamoog-modified.txt](https://github.com/csound/csound/files/1996959/csound-output-luamoog-modified.txt)

I have also noticed that if I remove the `lua_ikopcall` line but keep the `lua_opdef` command, Csound doesn't generate the "attempt to call a nil value" errors or hang, but it still produces a segmentation fault message. If I remove both the `lua_ikopcall` and `lua_opdef` commands from the example, Csound processes the file without errors.

**Comment by allen-marshall** [comment]
I tried creating a simpler Csound file using a Lua opcode that does nothing. I still get the Lua errors and segmentation fault, but Csound doesn't hang with the minimal example for some reason.

[Here](https://github.com/csound/csound/files/1997004/luasimple.zip) are the minimal example Csound file and the output I get from Csound.

**Comment by gogins** [comment]
I am investigating this. I can reproduce your error. I have debugged the
code and found that the code for running Lua opcodes no longer correctly
type casts between Lua and ctypes. Most likely there has either been some
subtle change in LuaJIT, or some change in the way Csound compiles opcodes
with many blank arguments. I will continue to look at this tomorrow.

Regards,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by gogins** [comment]
Some progress on this. My new code for managing Lua states per thread isn't
working. Should be easy to fix.


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by gogins** [comment]
I have fixed this in the develop branch. The bug was caused by my incorrectly assuming that `csoundGetCurrentThreadId()` returns the ID of a thread when actually it returns a pointer that must be dereferenced to obtain the ID of the thread, which also differs by operating system. Thus a new lua_State object was created every time a lua_State was requested, even from the same thread.

I fixed this by using `std::this_thread::get_id()` instead, which both functions correctly as a map key and is cross-platform.

Tested on Ubuntu 17.1
[luasimple.zip](https://github.com/csound/csound/files/1998731/luasimple.zip)
0 with the attached files.
